src/main/java/org/opensearch/neuralsearch/search/util/HybridSearchSortUtil.java

```diff
@@ -0,0 +1,116 @@
+        }
+        for (TopDocs topDoc : topDocs) {
+            if (topDoc != null && topDoc instanceof TopFieldDocs) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Creates a Sort object based on the provided top field documents.
+     * This method takes an array of top field documents and processes each field to create a corresponding SortField.
+     * The created  SortField objects are then used to create and return a new Sort object.
+     * @param topFieldDocs array of top field documents need to be processed.
+     * @return Sort object created based on provided top field documents.
+     */
+    private static Sort createSort(TopFieldDocs[] topFieldDocs) {
+        if (topFieldDocs == null || topFieldDocs[0] == null) {
+            throw new IllegalArgumentException("topFieldDocs cannot be null when sorting is applied");
+        }
+        final SortField[] firstTopDocFields = topFieldDocs[0].fields;
+        final SortField[] newFields = new SortField[firstTopDocFields.length];
+
+        for (int i = 0; i < firstTopDocFields.length; i++) {
+            final SortField delegate = firstTopDocFields[i];
+            final SortField.Type type = delegate instanceof SortedNumericSortField
+                ? ((SortedNumericSortField) delegate).getNumericType()
+                : delegate.getType();
+
+            if (SortedWiderNumericSortField.isTypeSupported(type) && isSortWideningRequired(topFieldDocs, i)) {
+                newFields[i] = new SortedWiderNumericSortField(delegate.getField(), type, delegate.getReverse());
+            } else {
+                newFields[i] = firstTopDocFields[i];
+            }
+        }
+        return new Sort(newFields);
+    }
+
+    /**
+     * Checks if sort widening is required for the provided top field documents.
+     *
+     * This method iterates through the provided topFieldDocs array and checks if any adjacent pairs of sort fields at the specified index are not equal.
+     * If any such pair is found, it returns true, indicating that sort widening is required. Otherwise, it returns false.
+     * @param topFieldDocs array of top field documents to be checked.
+     * @param sortFieldindex index of the sort field to be checked within each top field document.
+     * @return true if sort widening is required, false otherwise.
+     */
+    private static boolean isSortWideningRequired(TopFieldDocs[] topFieldDocs, int sortFieldindex) {
+        for (int i = 0; i < topFieldDocs.length - 1; i++) {
+            TopFieldDocs currentTopFieldDoc = topFieldDocs[i];
+            TopFieldDocs nextTopFieldDoc = topFieldDocs[i + 1];
+            if (currentTopFieldDoc == null || nextTopFieldDoc == null) {
+                throw new IllegalArgumentException("topFieldDocs cannot be null when sorting is applied");
+            }
+            if (!currentTopFieldDoc.fields[sortFieldindex].equals(nextTopFieldDoc.fields[sortFieldindex])) {
```

Comment:
Is it check reliable? I check for TopDocs it's not overridden, so we're using default equals which essentially checking for the same ref in memory. Should we check based on the object fields instead?

Reply:
Yes this check is reliable. The same is done in OpenSearch core. [https://github.com/opensearch-project/OpenSearch/blob/8ad199dac020c91c80180b4f8042f65c9994a81a/server/src/main/java/org/opensearch/action/search/SearchPhaseController.java#L630](https://github.com/opensearch-project/OpenSearch/blob/8ad199dac020c91c80180b4f8042f65c9994a81a/server/src/main/java/org/opensearch/action/search/SearchPhaseController.java#L630)